File: Classes/TiCaptureSession.m
//
//  TiCaptureSession.m
//  titanium-coreml
//
//  Created by Hans Knöchel on 05.07.17.
//

#import "TiCaptureSession.h"
#import "TiApp.h"

@implementation TiCaptureSession

- (instancetype)initWithCompletionHandler:(TiCaptureSessionCompletionHandler)completionHandler
{
    if (self = [super init]) {
        _completionHandler = completionHandler;
        _queue = dispatch_queue_create("com.axway.ti.coreml.camera-queue", NULL);
        [self setupCaptureSession];
    }
    
    return self;
}

- (void)setupCaptureSession
{
    _captureSession = [[AVCaptureSession alloc] init];
    _videoOutput = [[AVCaptureVideoDataOutput alloc] init];
    
    [[self captureSession] beginConfiguration];
    
    [[self captureSession] setSessionPreset:AVCaptureSessionPresetMedium];
    
    AVCaptureDevice *captureDevice = [AVCaptureDevice defaultDeviceWithMediaType:AVMediaTypeVideo];
    NSError *videoInputError = nil;
    
    if (captureDevice == nil) {
        NSLog(@"Error: Could not create capture device");
    }
    
    AVCaptureDeviceInput *videoInput = [[AVCaptureDeviceInput alloc] initWithDevice:captureDevice error:&videoInputError];
    
    if (videoInputError != nil) {
        NSLog(@"Error: Could not create video input");
    }
    
    if ([self.captureSession canAddInput:videoInput]) {
        [[self captureSession] addInput:videoInput];
    } else {
        NSLog(@"Error: Cannot add video input to capture session");
    }
    
    _previewLayer = [[AVCaptureVideoPreviewLayer alloc] initWithSession:self.captureSession];
    _previewLayer.videoGravity = AVLayerVideoGravityResizeAspect;
    _previewLayer.connection.videoOrientation = AVCaptureVideoOrientationPortrait;
    
    NSDictionary *settings = @{
        (id)kCVPixelBufferPixelFormatTypeKey: [NSNumber numberWithUnsignedInt:kCVPixelFormatType_32BGRA]
    };
    
    _videoOutput.videoSettings = settings;
    _videoOutput.alwaysDiscardsLateVideoFrames = YES;
    [_videoOutput setSampleBufferDelegate:self queue:_queue];
    
    if ([[self captureSession] canAddOutput:_videoOutput]) {
        [[self captureSession] addOutput:_videoOutput];
    } else {
        NSLog(@"Error: Cannot add video output to capture session");
    }
    
    [_videoOutput connectionWithMediaType:AVMediaTypeVideo];
    
    [self.captureSession commitConfiguration];
}

- (void)start
{
    [[self captureSession] startRunning];
}

- (void)stop
{
    [[self captureSession] stopRunning];
}

// MARK: AVCaptureVideoDataOutputSampleBufferDelegate

- (void)captureOutput:(AVCaptureOutput *)output didOutputSampleBuffer:(CMSampleBufferRef)sampleBuffer fromConnection:(AVCaptureConnection *)connection
{
    _completionHandler(CMSampleBufferGetImageBuffer(sampleBuffer));
}

@end

File: Classes/TiCaptureSession.swift
//
//  TiCaptureSession.swift
//  TiCoreml
//
//  Created by Hans Knöchel on 25.02.24.
//

import UIKit
import AVFoundation
import CoreVideo

typealias TiCaptureSessionCompletionHandler = (CVImageBuffer) -> Void

class TiCaptureSession: NSObject {
  
  let queue = DispatchQueue(label: "io.tidev.coreml.camera-queue")
  let completionHandler: TiCaptureSessionCompletionHandler
  let captureSession = AVCaptureSession()
  let videoOutput = AVCaptureVideoDataOutput()

  var previewLayer: AVCaptureVideoPreviewLayer!
  
  init(completionHandler: @escaping TiCaptureSessionCompletionHandler) {
    self.completionHandler = completionHandler

    super.init()

    self.setupCaptureSession()
  }
  
  private func setupCaptureSession() {
    captureSession.beginConfiguration()
    captureSession.sessionPreset = .photo
    
    guard let captureDevice = AVCaptureDevice.default(for: .video) else {
      return logErrorAndFail("Could not create capture device")
    }
    
    guard let videoInput = try? AVCaptureDeviceInput(device: captureDevice) else {
      return logErrorAndFail("Could not create video input")
    }
    
    guard captureSession.canAddInput(videoInput) else {
      return logErrorAndFail("Could not add video input to capture session")
    }
    
    captureSession.addInput(videoInput)
    
    previewLayer = AVCaptureVideoPreviewLayer(session: captureSession)
    previewLayer.videoGravity = .resizeAspect

//    if #available(iOS 17.0, *) {
//      previewLayer.connection?.videoRotationAngle = 90
//    } else {
      previewLayer.connection?.videoOrientation = .portrait
//    }
    
    videoOutput.videoSettings = [kCVPixelBufferPixelFormatTypeKey as String: kCVPixelFormatType_32BGRA]
    videoOutput.alwaysDiscardsLateVideoFrames = true
    videoOutput.setSampleBufferDelegate(self, queue: queue)
    
    guard captureSession.canAddOutput(videoOutput) else {
      return logErrorAndFail("Could not add video output to capture session")
    }
    
    captureSession.addOutput(videoOutput)

    let _ = videoOutput.connection(with: .video)
    captureSession.commitConfiguration()
  }
  
  func start() {
    captureSession.startRunning()
  }
  
  func stop() {
    captureSession.stopRunning()
  }
}

// MARK: AVCaptureVideoDataOutputSampleBufferDelegate

extension TiCaptureSession: AVCaptureVideoDataOutputSampleBufferDelegate {

  func captureOutput(_ output: AVCaptureOutput, didOutput sampleBuffer: CMSampleBuffer, from connection: AVCaptureConnection) {
    if let image = CMSampleBufferGetImageBuffer(sampleBuffer) {
      completionHandler(image)
    }
  }
}
